subsys/senss/include/senss_sensor.h

```diff
@@ -0,0 +1,605 @@
+ *
+ * Unblocked function, returned immediately.
+ *
+ * Used by a virtual sensor to post data to it's clients.
+ *
+ * A reporter sensor can use this API to post data to it's clients.
+ * For example, when a virtual sensor computed a data, then can use this API
+ * to deliver the data to it's clients.
+ * Pleas note, this API just for reporter post data to the sensor subsystem
+ * runtime, the runtime will help delivered the data to it's all clients
+ * according clients' configurations such as reporter interval, data change sensitivity.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param buf The data buffer.
+ *
+ * @param size The buffer size in bytes.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_sensor_post_data(
+		const struct device *dev,
+		void *buf, int size);
+
+/**
+ * @brief Get reporter handles	of a given sensor instance by sensor type.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param type The given type, \ref SENSS_SENSOR_TYPE_ALL to get reporters
+ * with all types.
+ *
+ * @param max_handles The max count of the \p reporter_handles array input. Can
+ * get real count number via \ref senss_sensor_get_reporters_count
+ *
+ * @param *reporter_handles Input handles array for receiving found reporter
+ * sensor instances
+ *
+ * @return number of reporters found, 0 returned if not found.
+ */
+int senss_sensor_get_reporters(
+		const struct device *dev, int type,
+		const senss_sensor_handle_t *reporter_handles, int max_handles);
+
+/**
+ * @brief Get reporters count of a given sensor instance by sensor type.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param type The sensor type for checking, \ref SENSS_SENSOR_TYPE_ALL
+ *
+ * @return Count of reporters by \p type, 0 returned if no reporters by \p type.
+ */
+int senss_sensor_get_reporters_count(
+		const struct device *dev, int type);
+
+/**
+ * @brief Get this sensor's state
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param state Returned sensor state value
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_sensor_get_state(
+		const struct device *dev,
+		enum senss_sensor_state *state);
+
+
+/**
+ * @}
+ */
+
+/**
+ * @brief Sensor Callbacks
+ * @addtogroup senss_sensor_callbacks
+ * \{
+ */
+
+/**
+ * @brief Sensor initialize callback for virtual sensor.
+ *
+ * Sensor can initialize it's runtime context in this callback.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param info The sensor instance's constant information.
+ *
+ * @param reporter_handles The reporters handles for this sensor.
+ *
+ * @param reporter_count The number of reporters.
+ *
+ * @return 0 on success or negative error value on failure.
+ *
+ */
+typedef int (*senss_sensor_virtual_init_t)(
+		const struct device *dev, const struct senss_sensor_info *info,
+		const senss_sensor_handle_t *reporter_handles, int reporters_count);
```

Comment:
Discussed in a meeting so I'm adding the comment here for visibility.

I don't believe that the sensor subsystem should be parsing the DTS bindings for the sensor. It's too limiting. Instead the init function should be simpler. It should tell the sensor which instance is being initialized (`const struct device *`) and that's it.

During the init, the virtual sensor can simply register itself as a listener via `sensor_subsystem_register_callback()` as I mentioned above which generates the linked list of callback objects. The sensor can then make requests to the subsystem to get specific data events and/or send configuration requests.

Reply:
> I think the core concept of `senss` is the [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) (directed acyclic graph) computing framework, it's power full, flexible and natural for sensor fusion.
> 
> In the DAG, physical sensors are data producers and focus on how to get data, read data from underlying sensor device, read data from wifi etc, virtual sensors are data consumers and producers, it's focus on fusion new data with reporters input data (physical sensors or virtual sensors). It's `senss`'s responsibility to connect with them and handle the arbitration on multiple clients report relationships between them.
> 
> DTS help config and describe the whole DAG, `senss` parse the DTS and build a runtime DAG.
> 
> And for a virtual sensor, all it's need are it's reporters, config each reporter, get reporter's information, receive reporter's data etc. Now the `init` give the virtual sensor instance 's reporters handles, with these handles, virtual sensor can know the reporters information, config them with needed configurations, then `process` callback give the reporters sensor data, virtual sensor can compute out new fusion sensor data with them and post to `senss`, `senss` help deliver to its clients (app or other virtual sensors).
> 
> Although, virtual sensor can use a listener via `sensor_subsystem_register_callback()`, but how does it to know it's reporters ? Parse DTS by itself ? All virtual sensors need to do this ? I think, its more better for `senss` help to do this, this is the value of the framework.

I believe that most if not all virtual sensors will need to parse the DTS themselves to get configurations custom to their binding. In the same manner, a lid angle should parse the DTS and make sure that it has exactly 2 phandles for the accelerometers. This can be done statically which is significantly better than having the subsystem account for every sensor (especially since we will likely allow 3rd party types to be added downstream).

> 
> Now with `SENSS_SENSOR_DT_DEFINE` to define a sensor (physical or virtual), it's based on `struct device` and device tree, all is determined at compile time, should have no the issue you mentioned in `sensor_subsystem_register_callback()`
> 
> Physical sensors now have the same `init` callback, the only difference is , `senss` will pass empty reporter information to a physical sensor, in fact, have no reporters is the only difference of a physical sensor compared to virtual sensor, this is also aligned with DTS definition. And now not limited physical sensor to read data from underlying sensor device, it's can get data in anyway.
> 
> So, I don't think its a limitation here.

I agree that it's not a limitation, I'm just not a fan of having 2 code paths that do the same thing. Here you have the ability to:
1. Get the reporters at runtime (in the init function)
2. Configure the reporters (which I'm not seeing a client handle here so that will need to be addressed)
3. Get your data using a different approach than how clients would

Above, # 2 is a concern (at least as far as I can tell) because you'll still need to arbitrate. This means that the virtual lid angle requesting the 2 accelerometer configurations needs to be treated just like the application requesting those accelerometers to be configured. It will need to have a client handle just like the application. So why have a different code path for it?

Reply:
For # 2, the reporters' `handles` will be passed to the virtual sensor via it's `init` callback,  virtual sensor can use these `handles` to configure the reporter sensors using `senss_set_xxx` just like Applications. And the subsystem will handle the multiple client's arbitration. 

Just as the discussion in https://github.com/zephyrproject-rtos/zephyr/pull/45370#discussion_r901253553, internally the subsystem treat Application clients and virtual sensor clients almost the same, and reused the single core logic and path. 

The difference is Application need explicit call API `senss_open_sensor()` to create clients to sensor instance(s), but for virtual sensors, the subsystem will auto create the clients to its reporters based on `devicetree`. After the clients are created, either from Application or Virtual sensors, internally the sussystem will treat them almost the same.  

From a subsystem's view, we don't know which Applications will use this susbsystem, and which sensors they will accessed, so we need define APIs such as `senss_open_sensor()` and `senss_close_sensor()` for Applications dynamic usages. But as subsystem itself, we can define and configure how many and which types of sensor instances, and how their reporting relationships, the best way to define and configure them is via `devicetree` and `kconfig`. 

From Application's view, they only need care about which types of sensors does the subsystem support and access the sensors they needed, not care about (and not aware of) the sensors is virtual or physical (just subsystem internal concepts for help create fusion sensors, such as hinge angle sensor).   

Future more, Application clients are just pure data consumers, virtual sensors are both data consumers and new sensor data producers (sensor fusion), so formally they will have different callbacks, but internally they still reused the same core logics and path in the subsystem.